### ISSUE TITLE
Enhancement idea: A different location proposal for administrators seems sensible

### ISSUE THREAD

**Issue by codemanyak** [fesch/Upla#2]
If it is possible to configure the setup in a way that it may distinguish administrative permissions from normal user permissions then it should be considered to propose a more central target location with common access rather than a user-specific AppData subfolder. If this automatic detection is not viable (not possible, too expensive or simply unwanted) then perhaps an additional setup step or checkbox might be inserted to decide whether a common installation is preferred, and in this case a location in either a system directory or in the common user subdirectory might be proposed. Likewise, the shortcuts should then be placed in the common environment rather than in the user-account-specific area.
The proposed target directory may not have common writing access in itself but then the administrative installer MUST launch the application after the installation has finished in order to establish the privileged directory entry for the downloaded file, which must sensibly be equipped with common write access such that any user may induce an update when launching the installed application (which in turn may involve some security risks, of course, not only if the URL or DNS gets compromised but also by means of direct overwriting...).

**Comment by fesch** [comment]
I'll check to see what the Inno Installer is capable of doing ...

**Comment by fesch** [comment]
I have uploaded a new installer, so you may test. Now the user has all choices ...

**Comment by codemanyak** [comment]
Looks better now at the beginning (I'm fine with the question whether to install it for only the current user or all users) and sensibly proposes the respective folder but then fails to download the Jar file when installed to the (x86) system programs folder:
![PathNotFound](https://user-images.githubusercontent.com/15326471/62058031-ff187000-b220-11e9-9723-05da25bb3e62.png)

My first guess is that the Java or system operations don't cope with the URI representation of the path (blank coding). I will do some tests in Eclipse to clarify this.

Apart from fixed issue #1, the handling of exceptions isn't much better in other cases, e.g. after the above error message the process remains stalled with splash "Downloading...":
![Launcher_06](https://user-images.githubusercontent.com/15326471/62058346-94b3ff80-b221-11e9-98ac-6e1b560ebdda.png)
Without a network connection, the Process likewise gets stuck with the splash screen "Testing network..." (also after the error message had popped up).
(So possibly UpLa main is meant always to exit in case of an exception that cannot be mended, otherwise the Launcher thread seems to remain in some zombie state.)

**Comment by fesch** [comment]
Without a network connection, Upla should either start the locally available JAR file or, if no JAR file is present, pop up the following dialog:
![image](https://user-images.githubusercontent.com/1027354/62066751-583ccf80-b232-11e9-8f7d-ae1d6d5f13ec.png)

I've tested this by disabling all my network connections.

Just notice, that testing if a network is available could take a couple of moments .... Consecutive retries will be faster (at least on Windows, I suppose the OS doing some caching)

**Comment by fesch** [comment]
Path bugs (Windows & Linux) have been removed in the latest push. The system now also exists with status code #1 if any exception occurs.

The installer on the site has been updated and I have tested it on a different computer.

If installed into the common application folder, the administrator must give write access to users on the directory of at least on the "structorizer.jar" in order to make everything work.

**Comment by fesch** [comment]
Inno setup allows to set permissions, so I succeeded in modiyfing the setup such that "structorizer.jar" get full permissions on install ;)

**Comment by codemanyak** [comment]
Just tested the newest version and can't find an obvious flaw anymore. Well done!

**Comment by fesch** [comment]
Great!

I've also succeed in signing the setup.exe, but my Windows still says that it comes from an unknown source although I use an official certificate :-( 

At least the signing works ;)

**Comment by codemanyak** [comment]
W.r.t. signing an executable for Windows, you are definitively not a newbie but it is always al little tricky.  Perhaps the following article may give some additional hint: [Code Signing a Windows application](https://mkaz.blog/code/code-signing-a-windows-application/).

Just a different question: In the program registry, the installation is listed as "Structorizer version 3.29-11" (and now properly offered for uninstallation). I wonder whether the version number would change on downloading an update.

**Comment by fesch** [comment]
Thanks for the link.

No, the version number won't change when a new update hat been dynamically downloaded as the Inno installer has nothing to do directly with Upla. Changing this also dynamically would mean Upla or Structorizer itself would need to deal with doing changes in the Windows registry ... and that's no fun ...

**Comment by codemanyak** [comment]
That's what I thought. Well, in this case you might at least try to avoid the integration of the version number in the product name - it is redundant anyway (see the following screenshot):
![SystemControl_Programs_1](https://user-images.githubusercontent.com/15326471/62108722-0b490f80-b2ab-11e9-8ba4-b2cc6da9825d.png)

The orange box referes to another aspect that came to my mind (though requiring more efforts): The setup might possibly support a configuration how updates are installed: automatically without consulting the user or only after a prompt allowing the user either to agree or to reject/postpone. Some users might prefer to have the opportunity to postpone or skip an update.  Product installations may specify that uninstalling isn't the only possible effect when calling the setup again, e.g. (see the orange boxes):
![SystemControl_Programs_2](https://user-images.githubusercontent.com/15326471/62109291-3e3fd300-b2ac-11e9-8db9-ca08b3947d66.png)
So the setup might offer to modify the update mode at any time or to uninstall.
EDIT: Maybe the update mode can be integrated in the UpLa.ini and implemented in the UpLa itself. (This would of course require that the setup is able to manipulate the UpLa.ini file). The good question, though, is how to inform about the features of the new version in a product-independent way - maybe just by presenting the link to the product homepage?

**Comment by codemanyak** [comment]
Maybe the following Inno Setup option allows to add an additional configurator that might handle the update mode for UpLa?
![InnoSetup_AppModifyPath](https://user-images.githubusercontent.com/15326471/62119942-f0819580-b2c0-11e9-84c1-cba9c6eb0978.png)
As a consequence, a separate "Modify" button is to be expected like in the following example:
![programs_and_features_4](https://user-images.githubusercontent.com/15326471/62120400-d09ea180-b2c1-11e9-99ef-233b51cf0ecf.png)

**Comment by fesch** [comment]
I succeeded in removing the redundant version from the name.

I also managed to modify the installer so that the "modify" (Ändern) button gets displayed and calls Upla with some special parameter (Upla not yet doing more than popping up some message right now).

The installer now also creates the links to the .nsd, .arr and .arrz files.

**Comment by codemanyak** [comment]
> I also managed to modify the installer so that the "modify" (Ändern) button gets displayed and calls Upla with some special parameter (Upla not yet doing more than popping up some message right now).

This is good news. But the question is whether Upla will be called with or can acquire sufficient rights to modify the upla.ini, which should still require administrative rights if placed in the OS programs folder. As far as I know the UAC can be provoked via the manifest in a Launch4J environment. On the other hand, this should be avoided if installed in a user subdirectory (account-specific installation). So I wonder if the manifest can contain conditional requirements or how the Installer could be made installing different files in the two different installation paradigms.

**Comment by fesch** [comment]
Sorry, I don't get the point here :(

If the installation will be done into the users folder, the latter has the rights to change the file anyway.

If the installation will be done into the common application folder, calling the "modify" will elevate into administrator mode anyway and thus will require administrative privileges, which then can change the INI file without problem.

**Comment by codemanyak** [comment]
> If the installation will be done into the common application folder, calling the "modify" will elevate into administrator mode anyway and thus will require administrative privileges, which then can change the INI file without problem.

This answers my question indeed. I wasn't quite sure whether the privileges will be elevated automatically or not. Fine if they will. (I hoped so.)
